### DISCUSSION TITLE
UDF Plugin: Which is the corresponding Java type to map(string,string)

### DISCUSSION THREAD

**Discussion by andreclaudino** [trinodb/trino#14096] [Q&A]
I am creating a plugin to host an UDF, it is an http_get function, receiving the `httpAddress` path, the query `parameters` and the request `headers`. It is implemeted in scala:

```SCALA
  @ScalarFunction(value = "http_get", deterministic = true)
  @Description("Returns the result of an Http Get request")
  @SqlType(value = StandardTypes.VARCHAR)
  def httpGetFromArrayMap(
                           @SqlType(StandardTypes.VARCHAR) httpAddress: Slice,
                           @SqlType(constants.STRING_MAP) parameters: ImmutableMap[Slice, Slice],
                           @SqlNullable @SqlType(constants.STRING_MAP) headers: ImmutableMap[Slice, Slice],
                         ): String = {

    val stringHeaders = castSliceMap(headers)
    val stringParams = castSliceMap(parameters)
    val request = Http(httpAddress.toStringUtf8).headers(stringHeaders).params(stringParams)
    val stringResponse = request.asString.body
    stringResponse
  }
```

When running on trino, it raises the following exception:

```
io.trino.spi.TrinoException: Exact implementation of http_get do not match expected java types.
```

The problem is: What is the corresponding java type to `map(varchar,varchar)`?

I've tried Many:

* Scala Map[String, String]
* Java Map<String, String>
* Java Map<Slice, Slice>
* ImmutableMap<String, String>
* ImmutableMap<Slice, Slice>

Can't find any example of plugin implementing a function which receives a map.

Any help is appreciated. Thanks

**Comment by ebyhr** [comment]
`Map` should be `Block`. I think `MapEntriesFunction` helps you as the example. 

By the way, you can ask questions about development in https://trino.io/slack.html `#dev` channel.

**Comment by nineinchnick** [comment]
This is now documented, with examples how to build such block https://trino.io/docs/current/develop/connectors.html#type-mapping
